pinot-core/src/main/java/com/linkedin/pinot/core/common/DataFetcher.java

```diff
@@ -0,0 +1,190 @@
+   *
+   * @param column column name.
+   * @return data source associated with this column.
+   */
+  public DataSource getDataSourceForColumn(String column) {
+    DataSource dataSource = _columnToDataSourceMap.get(column);
+    if (dataSource == null) {
+      dataSource = _indexSegment.getDataSource(column);
+      _columnToDataSourceMap.put(column, dataSource);
+    }
+    return dataSource;
+  }
+
+  /**
+   * Given a column, fetch its dictionary.
+   *
+   * @param column column name.
+   * @return dictionary associated with this column.
+   */
+  public Dictionary getDictionaryForColumn(String column) {
+    Dictionary dictionary = _columnToDictionaryMap.get(column);
+    if (dictionary == null) {
+      dictionary = getDataSourceForColumn(column).getDictionary();
+      _columnToDictionaryMap.put(column, dictionary);
+    }
+    return dictionary;
+  }
+
+  /**
+   * Given a column, fetch its block value set.
+   *
+   * @param column column name.
+   * @return block value set associated with this column.
+   */
+  public BlockValSet getBlockValSetForColumn(String column) {
+    BlockValSet blockValSet = _columnToBlockValSetMap.get(column);
+    if (blockValSet == null) {
+      blockValSet = getDataSourceForColumn(column).getNextBlock().getBlockValueSet();
+      _columnToBlockValSetMap.put(column, blockValSet);
+    }
+    return blockValSet;
+  }
+
+  /**
+   * Init the data fetcher with a block doc id array and block length. This method should be called before fetching data
+   * for any specific block.
+   *
+   * @param blockDocIdArray block doc id array.
+   * @param blockLength length of the block.
+   */
+  public void initNewBlock(int[] blockDocIdArray, int blockLength) {
+    _columnDictIdLoaded.clear();
+    _columnValueLoaded.clear();
+    _columnHashCodeLoaded.clear();
+
+    _blockDocIdArray = blockDocIdArray;
+    _blockLength = blockLength;
+  }
+
+  /**
+   * Get dictionary id array for a given column for the specific block initialized in the initNewBlock.
+   *
+   * @param column column name.
+   * @return dictionary id array associated with this column.
+   */
+  public int[] getDictIdArrayForColumn(String column) {
+    int[] dictIdArray = _columnToDictIdArrayMap.get(column);
+    if (!_columnDictIdLoaded.contains(column)) {
+      if (dictIdArray == null) {
+        dictIdArray = new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+        _columnToDictIdArrayMap.put(column, dictIdArray);
+      }
+      BlockValSet blockValSet = getBlockValSetForColumn(column);
+      blockValSet.readIntValues(_blockDocIdArray, 0, _blockLength, dictIdArray, 0);
+      _columnDictIdLoaded.add(column);
+    }
+    return dictIdArray;
+  }
+
+  /**
+   * Get value array for a given column for the specific block initialized in the initNewBlock.
+   *
+   * @param column column name.
+   * @return value array associated with this column.
+   */
+  public double[] getValueArrayForColumn(String column) {
```

Comment:
without support for handling all data types and single value/multi value, this class is not adding much value.